File: test/language/templates/cpp/ChoiceTemplatedFieldTest.cpp
#include "gtest/gtest.h"
#include "templates/choice_templated_field/ChoiceTemplatedField.h"

namespace templates
{
namespace choice_templated_field
{

TEST(ChoiceTemplatedFieldTest, readWrite)
{
    ChoiceTemplatedField choiceTemplatedField;
    choiceTemplatedField.setSelector(0);
    {
        TemplatedChoice_uint32_uint16 choice1;
        choice1.setTemplatedField1(42);
        choiceTemplatedField.setChoice1(std::move(choice1));
    }
    {
        TemplatedChoice_Compound_uint32_uint16 choice2;
        choice2.setTemplatedField1(Compound_uint32{42});
        choiceTemplatedField.setChoice2(choice2); // copy
    }
    choiceTemplatedField.initializeChildren();

    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);
    zserio::BitStreamWriter writer(bitBuffer);
    choiceTemplatedField.write(writer);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    ChoiceTemplatedField readChoiceTemplatedField(reader);

    ASSERT_TRUE(choiceTemplatedField == readChoiceTemplatedField);
}

} // namespace choice_templated_field
} // namespace templates

File: test/language/templates/python/ChoiceTemplatedFieldTest.py
import zserio

import Templates

class ChoiceTemplatedFieldTest(Templates.TestCase):
    def testReadWrite(self):
        selector = 0
        choice1 = self.api.TemplatedChoice_uint32_uint16(selector)
        choice1.templated_field1 = 42
        choice2 = self.api.TemplatedChoice_Compound_uint32_uint16(
            selector, templated_field1_=self.api.Compound_uint32(42)
        )
        choiceTemplatedField = self.api.ChoiceTemplatedField(selector, choice1, choice2)

        writer = zserio.BitStreamWriter()
        choiceTemplatedField.write(writer)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readChoiceTemplatedField = self.api.ChoiceTemplatedField()
        readChoiceTemplatedField.read(reader)
        self.assertEqual(choiceTemplatedField, readChoiceTemplatedField)
